[admission][auto-instrumentation] Support .NET

zacharycmontoya opened the pull request:
### What does this PR do?
This PR sets the required environment variables to automatically instrument a .NET application with the .NET Tracer using the Admission Controller.

### Motivation

### Additional Notes
The E2E test will need to be updated when the official .NET init image is published to GCR. For now, it's pinned to a dev build.

### Possible Drawbacks / Trade-offs

### Describe how to test/QA your changes
Deploy the cluster agent and auto-instrument a .NET app, validate tracing and continuous profiler is working e2e.

### Reviewer's Checklist
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

pierotibou committed:
Initial implementation of the dotnet library injection

zacharycmontoya committed:
Add LD_PRELOAD environment variable which is required for installing the .NET Continuous Profiler on Linux

zacharycmontoya committed:
Add e2e test

zacharycmontoya committed:
Add reno

zacharycmontoya commented:
The CircleCI Pipeline GitHub Action has automatically failed, do I need to add some Circle app permission to my GitHub account?

andrewlock commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -220,18 +239,56 @@ func injectLibRequirements(pod *corev1.Pod, ctrName string, envKey string, envVa
+			return err4
+		}
+
+		err5 := setEnvVarWithFunc(&pod.Spec.Containers[i], dotnetProfilingLdPreloadKey, dotnetProfilingLdPreloadEnvValFunc)
+		if err5 != nil {
+			return err5
 		}
```
Need to add the profiler 
```suggestion
		err5 := setEnvVarWithFunc(&pod.Spec.Containers[i], dotnetProfilerHomeKey, dotnetProfilerHomeValue)
		if err5 != nil {
			return err5
		}

		err6 := setEnvVarWithFunc(&pod.Spec.Containers[i], dotnetProfilingLdPreloadKey, dotnetProfilingLdPreloadEnvValFunc)
		if err6 != nil {
			return err6
		}
```

zacharycmontoya replied:
DD_DOTNET_PROFILER_HOME is no longer required, so we're good here 😄

zacharycmontoya commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -40,6 +40,22 @@ const (
+	dotnetClrProfilerPathKey   = "CORECLR_PROFILER_PATH"
+	dotnetClrProfilerPathValue = "/datadog-lib/Datadog.Trace.ClrProfiler.Native.so"
+
+	dotnetTracerHomeKey   = "DD_DOTNET_TRACER_HOME"
+	dotnetTracerHomeValue = "/datadog-lib"
+
```
DD_DOTNET_PROFILER_HOME is no longer required, so we're good here 😄

andrewlock replied:
oooh, so it is! Unrelated, but I guess we should remove it from the MSI (and docs if it's used there) too then.

andrewlock replied:
Created a PR for the cleanup here (it's already removed from docs): https://github.com/DataDog/dd-trace-dotnet/pull/3782

vboulineau commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -220,18 +239,56 @@ func injectLibRequirements(pod *corev1.Pod, ctrName string, envKey string, envVa
+}
+
+// injectLibRequirements injects the minimal config requirements to enable instrumentation
+func injectLibRequirementsDotnet(pod *corev1.Pod, ctrName string) error {
```
It seems like something like:

```
type envVar struct {
  key string
  valFunc envValFunc
}

func injectLibRequirements(pod *corev1.Pod, ctrName string, envVars []enVar) error
```

Would be more suited

zacharycmontoya replied:
I could implement this, but I would also require creating a total of 5 new functions to set the .NET environment variables. Is that a trade-off you think is worth making?

zacharycmontoya replied:
Ah I see you're suggesting I compose these two suggestions together, ok can do 👍🏼 Thanks for the suggestions

zacharycmontoya replied:
Done in 491425ab90ce7dfdc556249806debafecddab177

vboulineau commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -220,18 +239,56 @@ func injectLibRequirements(pod *corev1.Pod, ctrName string, envKey string, envVa
+			return err
+		}
+
+		err2 := setEnvVar(&pod.Spec.Containers[i], dotnetClrProfilerIdKey, dotnetClrProfilerIdValue)
```
Creating new `errX` obj is not useful as we stop whenever we get one, re-using `err` is probably more appropriate.

zacharycmontoya replied:
With the identity function you suggested, I removed the custom dotnet function entirely so this is no longer a concern. Done in 491425ab90ce7dfdc556249806debafecddab177

vboulineau commented on pkg/clusteragent/admission/mutate/auto_instrumentation.go:
```diff
@@ -266,6 +323,46 @@ func injectLibConfig(pod *corev1.Pod, lang language) error {
+	return nil
+}
+
+func setEnvVar(ctr *corev1.Container, envKey string, envValue string) error {
```
`setEnvVar` and `setEnvVarFunc` are actually the same with an identity `envValFunc`:
```
func identityValFunc(s string) envValFunc {
  return func(string) string { return s }
}
```

zacharycmontoya replied:
Added in 491425ab90ce7dfdc556249806debafecddab177

zacharycmontoya committed:
PR Feedback:
- Create an envVar struct to hold key and valFunc pair
- Update injectLibRequirements to receive a slice of envVar structs and update each language invocation
- Update dotnet handling to use injectLibRequirements and scrap its custom handling and the newly introduced setEnvVar functions

zacharycmontoya committed:
Add environment variable DD_TRACE_LOG_DIRECTORY for .NET so we can reliably emit tracer logs regardless of the application container's filesystem settings

frankbolero commented:
Is there some timeframe for when this will be released, we are currently enrolling in Datadog and auto instrumentation for dotnet would be incredibly nice
to have instead of having to manually alter each image.

elee0625 commented:
Hi @frankbolero - we are currently waiting on our agent release cycle, and expect this to be released sometime in early to mid April. We appreciate your interest and patience!
